### ISSUE TITLE
HOWTO: Auto restart/upload on broken connections

### ISSUE THREAD

**Issue by mrspirytus** [Azure/azure-sdk-for-cpp#2688]
**Query/Question**
I just implemented a test app to transfer files using BlockBlobClient::UploadFrom().
Since I will be uploading files from test equipment over LTE connections on moving vehicles, connections might come and go away.
What is the best way to handle

1. Disconnects and restarts
1. Most importantly if possible do not upload the entire file again but just restart from the last uploaded point. Sometimes might be large. Restart upload from the last known point would be awesome.
Can someone point me to a detailed example/document on how to achieve this?

***Why is this not a Bug or a feature Request?***
A clear explanation of why is this not a bug or a feature request?

**Setup (please complete the following information if applicable):**
 - OS: Ubuntu 20.04 and 18.04
 - IDE : vim, Qt Creator
 - Version of the Library used: 12.0.1
 
 **Information Checklist**
 Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [x] Query Added
- [x] Setup information Added

**Comment by vhvb1989** [comment]
@mrspirytus thank you for reaching out.

I know Storage service supports Downloading blobs setting an offset.  We use that to recover from network-interruptions while downloading.  But for uploading, I don't think that's supported by default.  

@Jinming-Hu will be able to give you a better answer here. (so, adding him).

@Jinming-Hu , would using `appendBlob` would work as an strategy for uploading small chunks a at time?  This might help to deal with network interruptions ?

**Comment by vhvb1989** [comment]
Also, @mrspirytus , the SDK client will give you retry-support by default.

If we call `uploadFrom(...)` and network is broken in the middle of the request, by default, the SDK client will try to complete the request after 4 seconds again (hoping connection is back).  

You can define/setup the number of retry intents  and and delay between them by using the sdk client options (last optional arg for creating a client).  There is a field `Retry` in the options to set this up.

Setting the number of retries to a high number might help you deal with network interruptions.  However, if the uploading payload is huge, it might be hard to complete the uploading operation (since it is always re-started from the start on each retry)

**Comment by mrspirytus** [comment]
@Jinming-Hu @vhvb1989 So what are you saying uploadFrom() will block until transfer is completed? unless retry count is exhausted? Let me describe my case better :)

I am running on embedded Linux on vehicles collecting various vehicle information.
When a specific amount of data is collected (~20Mb file.. might be more or less in future)
The embedded unit might shut down at any point possibly interrupting
   a. Active transfer
   b. Retry cycle (network is down)

If a system shutdown is requested, is there a way to cancel safely active uploadFrom() from another (shutdown) thread? Regardless if it is in state a. or b.?

I am assuming when the system boots again and I will start transferring the same "incomplete file", should I expect a full transfer or it will check with the remote server and start from the "broken" point? 

Also, I noticed that I can specify how many concurrent chunks I want to transfer (I believe it is 5 looking at the source). In my case, since I do not want to overwhelm LTE connection, since there are other active "control" connections on our application server, I am thinking of limiting one chunk at a time and also limit bandwidth use.

1. Can U change how much maximum bandwidth can transfer take?
2. Can this bandwidth be changed dynamically (If I determine that as some point I can try to transfer as fast as possible?

Thank you

**Comment by vhvb1989** [comment]
> uploadFrom() will block until transfer is completed?

Yes, all operations from the sdk client are synchronous.  We we call `auto response = client.FooOperation(...)`, the current thread is blocked until the operation is completed or cancelled.
All the operations take a context as the last argument which can be used to cancel the operation at any given time in the future (from another thread). Or the context can be also created with a timeout value to cancel the operation if it did not complete after the specified time.

@Jinming-Hu  can tell how to control the chunk size and concurrency better than me :)

> If a system shutdown is requested, is there a way to cancel safely active uploadFrom() from another (shutdown) thread? Regardless if it is in state a. or b.?

Yes, if you have a shutdown callback that is called on shutdown, you can let that function to cancel the context used for the upload operation.  The SDK exposes a global on Azure::Core::Context:ApplicationContext.  This global context can be used to generate more child context from it.  Then, if at any moment you call .Cancel() on the global context, all its child-context will be cancelled.  On a high concurrency scenario, where we have multiple threads running different SDK client operations, this technique would allow to cancel everything at the same time.

Let me know if you would like more info on how to use the context component.  There are some tests on how to use context with deadLine or directly cancelling it here: https://github.com/Azure/azure-sdk-for-cpp/blob/main/sdk/core/azure-core/test/ut/context_test.cpp#L91

**Comment by vhvb1989** [comment]
> I am assuming when the system boots again and I will start transferring the same "incomplete file", should I expect a full transfer or it will check with the remote server and start from the "broken" point?

Once the operation is cancelled, the SDK won't save anything related to that operation. You would need to implement a way to control if your upload-request completed and how to resume the incompleted-ones.

Whenever the upload(..) operation returns to you with a success code, you can update/mutate your state.  But if the operation throws or return with an error code, you might want to handle the exception and either try again later or update the request (if there's an error on the request).

The upload operation will handle retrying the request at least 3 times by default (and only for some known fail-cases where retry makes sense. For instance, it won't retry on auth-error or on Server returning an error code with description for a bad request). The sdk client will retry for connection-lost/interrupted/closed, for example.

My advise would be to perform the upload for small file size to ensure the operation complete as fast as possible.  A ~1Mb upload will complete faster (~<1sec) than ~20Mb (~1sec) with LTE working at 20Mbps.

**Comment by Jinming-Hu** [comment]
@mrspirytus I'd suggest you use append blob and use relatively small chunk size, which means instead of appending 20MB in one request, you append 1MB at a time. Take a look at [AppendBlock](https://github.com/Azure/azure-sdk-for-cpp/blob/8db9c0b223da1c85f80e0ed058be84b931cbea9f/sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/append_blob_client.hpp#L140) operation. This operation does not provide any concurrent transfer support, so it won't overwhelm your LTE connection.

> 1. Can U change how much maximum bandwidth can transfer take?
> 2. Can this bandwidth be changed dynamically (If I determine that as some point I can try to transfer as fast as possible?

No, we don't support bandwidth control.
